components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultOIDCClaimsCallbackHandler.java

```diff
@@ -0,0 +1,628 @@
+        return requestMsgCtx.getAuthorizedUser().isFederatedUser();
+    }
+
+    private Map<ClaimMapping, String> getUserAttributesCachedAgainstAuthorizationCode(Object authorizationCode) {
+        Map<ClaimMapping, String> userAttributes = Collections.emptyMap();
+        if (authorizationCode != null) {
+            // get the cached user claims against the authorization code if any
+            userAttributes = getUserAttributesFromCacheUsingCode(authorizationCode.toString());
+        }
+        return userAttributes;
+    }
+
+    private Map<ClaimMapping, String> getUserAttributesCachedAgainstToken(String accessToken) {
+        Map<ClaimMapping, String> userAttributes = Collections.emptyMap();
+        if (accessToken != null) {
+            // get the user claims cached against the access token if any
+            userAttributes = getUserAttributesFromCacheUsingToken(accessToken);
+        }
+        return userAttributes;
+    }
+
+    private Map<String, Object> getUserClaims(OAuthAuthzReqMessageContext authzReqMessageContext)
+            throws OAuthSystemException {
+
+        Map<String, Object> claims = Collections.emptyMap();
+        Map<ClaimMapping, String> userAttributes =
+                getUserAttributesCachedAgainstToken(getAuthorizationCode(authzReqMessageContext));
+
+        if (isEmpty(userAttributes))
+            if (!isFederatedUser(authzReqMessageContext)) {
+                claims = getClaimsForLocalUser(authzReqMessageContext, claims);
+            } else {
+                claims = getClaimsMap(userAttributes);
+            }
+        else {
+            claims = getClaimsMap(userAttributes);
+        }
+
+        String spTenantDomain = authzReqMessageContext.getAuthorizationReqDTO().getTenantDomain();
+        return filterClaimsByScope(authzReqMessageContext.getApprovedScope(), spTenantDomain, claims);
+    }
+
+    private Map<String, Object> getClaimsForLocalUser(OAuthAuthzReqMessageContext authzReqMessageContext, Map<String, Object> claims) {
+        try {
+            claims = getClaimsFromUserStore(authzReqMessageContext);
+            if (log.isDebugEnabled()) {
+                log.debug("User attributes not found in cache. Trying to retrieve attribute for user " +
+                        authzReqMessageContext.getAuthorizationReqDTO().getUser());
+            }
+        } catch (UserStoreException | IdentityApplicationManagementException | IdentityException e) {
+            log.error("Error occurred while getting claims for user " +
+                    authzReqMessageContext.getAuthorizationReqDTO().getUser(), e);
+        }
+        return claims;
+    }
+
+    private boolean isFederatedUser(OAuthAuthzReqMessageContext authzReqMessageContext) {
+        return authzReqMessageContext.getAuthorizationReqDTO().getUser().isFederatedUser();
+    }
+
+    private String getAuthorizationCode(OAuthAuthzReqMessageContext authzReqMessageContext) {
```

Comment:
**getAuthorizationCode**: method name is invalid. Should be like **getAccessToken**